File: backend/src/services/pipelineService.js
/**
 * 多 Agent 流水线服务
 * 协调多个 AI Agent 的顺序执行
 */

const aiService = require('./aiService');
const historyModel = require('../models/history');

class PipelineService {
    /**
     * 执行完整的解题流程
     * @param {Object} input - 输入数据
     * @param {string} input.type - 输入类型: 'text' | 'image'
     * @param {string} input.content - 文本内容或 Base64 图片
     * @param {string} input.userId - 用户ID（可选）
     * @param {string} input.username - 用户名（可选）
     * @returns {Promise<Object>} 完整结果
     */
    async solveProblem(input) {
        const result = {
            success: true,
            data: {}
        };

        try {
            // Step 1: 图像识别（如果是图片输入）
            let problemText = '';
            if (input.type === 'image') {
                console.log('开始图像识别...');
                problemText = await aiService.recognizeImage(input.content);
                result.data.recognizedText = problemText;
                console.log('图像识别完成');
            } else {
                problemText = input.content;
                result.data.recognizedText = problemText;
            }

            // Step 2: 题目解析
            console.log('开始题目解析...');
            const parseResult = await aiService.parseProblem(problemText);
            result.data.parseResult = parseResult;
            console.log('题目解析完成:', parseResult.type, parseResult.difficulty);

            // Step 3: 生成解答
            console.log('开始生成解答...');
            const solution = await aiService.generateSolution(problemText, parseResult);
            result.data.solution = solution;
            console.log('解答生成完成');

            // Step 4: 保存历史记录（如果用户已登录）
            if (input.userId) {
                console.log('保存历史记录...');
                const historyRecord = historyModel.create({
                    userId: input.userId,
                    username: input.username,
                    question: problemText,
                    parseResult: parseResult,
                    solution: solution
                });
                result.data.historyId = historyRecord.id;
                console.log('历史记录已保存:', historyRecord.id);
            }

            return result;
        } catch (error) {
            console.error('解题流程失败:', error);
            return {
                success: false,
                error: error.message,
                data: result.data
            };
        }
    }

    /**
     * 仅执行图像识别
     * @param {string} imageBase64 - Base64 图片
     * @returns {Promise<Object>} 识别结果
     */
    async recognizeOnly(imageBase64) {
        try {
            const text = await aiService.recognizeImage(imageBase64);
            return {
                success: true,
                data: { text }
            };
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * 仅执行题目解析
     * @param {string} text - 题目文本
     * @returns {Promise<Object>} 解析结果
     */
    async parseOnly(text) {
        try {
            const parseResult = await aiService.parseProblem(text);
            return {
                success: true,
                data: parseResult
            };
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * 仅执行解答生成
     * @param {string} text - 题目文本
     * @param {Object} parseResult - 解析结果
     * @returns {Promise<Object>} 解答结果
     */
    async solveOnly(text, parseResult) {
        try {
            const solution = await aiService.generateSolution(text, parseResult);
            return {
                success: true,
                data: solution
            };
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * 流式生成解答
     * @param {string} text - 题目文本
     * @param {Object} parseResult - 解析结果
     * @param {Function} onData - 数据回调
     */
    async solveStream(text, parseResult, onData) {
        try {
            await aiService.generateSolutionStream(text, parseResult, onData);
        } catch (error) {
            onData(null, true, error.message);
        }
    }
}

module.exports = new PipelineService();

File: backend/app/services/pipeline_service.py
"""Pipeline service coordinating recognition, parsing and solving."""

from __future__ import annotations

from typing import Dict, Generator

from app.extensions import db
from app.models.history import History
from app.services.ai_service import ai_service


class PipelineService:
    def solve_problem(self, input_data: Dict) -> Dict:
        result = {"success": True, "data": {}}

        try:
            if input_data.get("type") == "image":
                problem_text = ai_service.recognize_image(input_data.get("content", ""))
            else:
                problem_text = str(input_data.get("content", ""))

            result["data"]["recognizedText"] = problem_text

            parse_result = ai_service.parse_problem(problem_text)
            result["data"]["parseResult"] = parse_result

            solution = ai_service.generate_solution(problem_text, parse_result)
            result["data"]["solution"] = solution

            user_id = input_data.get("userId")
            if user_id:
                history_record = History(
                    user_id=user_id,
                    username=input_data.get("username"),
                    question=problem_text,
                    parse_result=parse_result,
                    solution=solution,
                )
                db.session.add(history_record)
                db.session.commit()
                result["data"]["historyId"] = history_record.id

            return result
        except Exception as exc:  # noqa: BLE001
            db.session.rollback()
            return {
                "success": False,
                "error": str(exc),
                "data": result["data"],
            }

    def recognize_only(self, image_base64: str) -> Dict:
        try:
            text = ai_service.recognize_image(image_base64)
            return {"success": True, "data": {"text": text}}
        except Exception as exc:  # noqa: BLE001
            return {"success": False, "error": str(exc)}

    def parse_only(self, text: str) -> Dict:
        try:
            parse_result = ai_service.parse_problem(text)
            return {"success": True, "data": parse_result}
        except Exception as exc:  # noqa: BLE001
            return {"success": False, "error": str(exc)}

    def solve_only(self, text: str, parse_result: Dict) -> Dict:
        try:
            solution = ai_service.generate_solution(text, parse_result)
            return {"success": True, "data": solution}
        except Exception as exc:  # noqa: BLE001
            return {"success": False, "error": str(exc)}

    def solve_stream(self, text: str, parse_result: Dict) -> Generator[str, None, None]:
        yield from ai_service.generate_solution_stream(text, parse_result)


pipeline_service = PipelineService()
